### PR TITLE
Initial commit

### PR DESCRIPTION
**Pull request by krancour**
Signed-off-by: Kent Rancourt <kent.rancourt@microsoft.com>

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -0,0 +1,257 @@
...
+chart. Helm 3 has _experimental_ support for OCI registries. In the event that
+the Helm 3 dependency proves troublesome for users, or in the event that this
+experimental feature goes away, or isn't working like we'd hope, we will revisit
+this choice before going GA.
+
+To fetch the Brigade Docker Hub Gateway chart from the registry:
+
+```console
+  export HELM_EXPERIMENTAL_OCI=1
+  helm chart pull ghcr.io/brigadecore/brigade-dockerhub-gateway:v2.0.0-alpha.1
```
Comment (vdice) [inline README.md:69]: As this is a brand new gateway (not a rewrite of an existing one), do we still want to start versioning w/ `v2.0.0-*`?  Naturally, this versioning does align with Brigade v2.  I could also see it starting with the usual `v0.1.0-*`.  No real preference on my end, just wanted to check.

File: README.md [hunk 2]
```diff
@@ -0,0 +1,257 @@
...
+
+__Note:__ Docker Hub doesn't provide _any_ reasonable mechanism for
+authenticating to the endpoints to which events (webhooks) are sent. Due to
+this, the only viable approach to authentication is to include a token (a shared
+secret) in the webhook URL as depicted above. Users are cautioned that even with
+TLS in play, this is not _entirely_ secure because web servers, reverse proxies,
+and other infrastructure are apt to capture entire URLs, including query
+parameters, in their access logs. _If your threat model suggests this is an
+intolerable degree of risk, then do not use this gateway and, more generally, do
+not user Docker Hub webhooks._
```
Comment (vdice) [inline README.md:143]:
```suggestion
not use Docker Hub webhooks._
```

File: charts/brigade-dockerhub-gateway/values.yaml [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+replicas: 1
+
+## Host should be set to the public IP address or DNS hostname for this
+## gateway. Whenever possible, it should be set accurately for a variety of
+## reasons. If applicable, the value is used both in automatic certificate
+## generation and, also if applicable, the creation of ingress routing rules
+## (advanced). Keeping the default value or otherwise setting this
+## inaccurately may result in invalid routing rules or untrusted certificates
+## (untrusted beyond merely having been self-signed).
+host: cr-gateway.example.com
```
Comment (vdice) [inline charts/brigade-dockerhub-gateway/values.yaml:14]: nit: maybe update to `dockerhub-gateway.example.com`?

File: README.md [hunk 4]
```diff
@@ -0,0 +1,257 @@
...
+
+### 4. Create Webhooks
+
+In your browser, go to your Docker Hub repository for which you'd like to send
+webhooks to this gateway. From the tabs across the top of the page, select
+__Webhooks__. 
+
+* In the __Webhook name__ field, add a meaningful name for the webhook.
+
+* In the __Webhook URL__ field, use a value of the form `https://<DNS hostname or publicIP>/events?access_token=<url-encoded token>`.
```
Comment (vdice) [inline README.md:131]: This URL takes a url-encoded token whereas the `tokens` chart field appears to take non-encoded token values.  If this is correct, should we add guidance on how to url-encode a token from the `tokens` field for building the URL here?  (Or, if not how, maybe just a mention that this needs to be done.)

File: internal/webhooks/token_filter.go [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+	AddToken(string)
+	getHashedTokens() []string
+}
+
+// tokenFilterConfig encapsulates token filter configuration.
+type tokenFilterConfig struct {
+	hashedTokens []string
+}
+
+// NewTokenFilterConfig returns a initialized implementation of the
```
Comment (vdice) [inline internal/webhooks/token_filter.go:26]:
```suggestion
// NewTokenFilterConfig returns an initialized implementation of the
```

File: internal/webhooks/token_filter.go [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+		// authentication scheme AT ALL. The best we can possibly rely on is that
+		// Docker Hub users can add a query parameter containing a token to the
+		// URL they use when defining their webhooks. This is why we select the
+		// token from a query parameter instead of a header.
+		//
+		// Further note that even with TLS in play, this is not *entirely* secure
+		// because web servers, reverse proxies, and other infrastructure are apt
+		// to capture entire URLs, including query parameters, in their access logs.
+		//
+		// User facing documentation MUST make a not of these risks.
```
Comment (vdice) [inline internal/webhooks/token_filter.go:72]:
```suggestion
		// User facing documentation MUST make a note of these risks.
```

File: .brigade/brigade.ts [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+async function runSuite(event: Event): Promise<void> {
+  await new SerialGroup(
+    new ConcurrentGroup( // Basic tests
+      testUnitJob(event),
+      lintJob(event),
+      lintChartJob(event)
+    ),
+    buildJob(event)
+  ).run()
+  if (event.worker?.git?.ref == "v2") {
```
[the remark this answers is not in the document]
Comment (krancour) [reply .brigade/brigade.ts:101]: D'oh!

File: README.md (hunk 1 above)
Comment (krancour) [reply README.md:69]: It should have been v0.1.0 initially.

Comment (krancour) [reply README.md:69]: I'll fix.

File: charts/brigade-dockerhub-gateway/values.yaml (hunk 3 above)
Comment (krancour) [reply charts/brigade-dockerhub-gateway/values.yaml:14]: Yup. This is something I missed after realizing that I couldn't really make a single "container registry" gateway that supported Docker Hub _and_ ACR. Will fix.

File: README.md (hunk 4 above)
Comment (krancour) [reply README.md:131]: I'll add a note to values.yaml about this.

File: internal/webhooks/token_filter.go (hunk 6 above)
Comment (krancour) [reply internal/webhooks/token_filter.go:72]: This was supposed to be a TODO. This is mentioned in the README, so I am actually just going to delete this line.

**Comment by krancour** [comment]
Updated based on feedback.

**Review by vdice — APPROVED** [review]

**Pull request merged**